tensorflow/contrib/opt/python/training/adamax.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""AdaMax for TensorFlow."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from tensorflow.python.framework import ops
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import control_flow_ops
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import resource_variable_ops
+from tensorflow.python.ops import state_ops
+from tensorflow.python.training import adam
+from tensorflow.python.training import training_ops
+
+
+class AdaMaxOptimizer(adam.AdamOptimizer):
+  """Optimizer that implements the AdaMax algorithm.
+
+  See [Kingma et al., 2014](http://arxiv.org/abs/1412.6980)
```

Comment:
Also explain why you might want to use this optimizer instead of Adam, from the same paper.

Reply:
I think @dchatterjee172 has explained the feature request in #17104:
>>> It will be really nice to have a adamax implementation along with the regular adam optimization algorithm. Adamax superior than adam in certain cases, generally in models with word embedding.

Reply:
So can you add "Adamax is sometimes superior to adam, specially in models with embeddings" to the comment?

Reply:
Fine.